controller/src/main/java/io/pravega/controller/server/rpc/grpc/GRPCServerConfig.java

```diff
@@ -95,6 +95,14 @@
      */
     Integer getAccessTokenTTLInSeconds();
 
+    /**
+     * Indicates whether writes to internal streams are allowed with read permissions. If false, writes to internal
+     * streams will require write permissions too.
+     *
+     * @return Whether writes to internal streams are allowed with read permissions.
+     */
+    boolean isInternalWritesWithReadPermEnabled();
```

Comment:
One major problem I see with configuring it this way is that it will be difficult for a user to understand what to do about it. It is rather counterintuitive and the tradeoffs are not very clear. What would be the argument against always giving write access to the reader group synchronizer stream instead of adding this parameter?

Reply:
@fpj, I am not sure if you meant what you said, which is "always give write access" or you meant "require write access".

I don't think we can "give" write access automatically.   The reason why we want to be able to preserve the "you need READ_UPDATE to touch/mutate this ReaderGroup" behavior is for these types of deployments where:

* environments where low level granular "ahead of time" acl provisioning is not possible due to the fact that the resources, scopes, streams, reader groups, are not known ahead of time.  
* there is a conscious intent to apply ACLs at the scope level because it's easier to manage for the admin, who does not have to worry about new streams getting created and having to micro manage at that level.  Scopes are more viewed like "teams" or "tenants".

For that reason, heavy use of wild cards is required usually at the scope level.  That means that if "READ" allowed access to reader groups in general, reader subjects could be interfering with each other.

An example of this is when members of a team called "**compute1**" scope for which was created so they can run their own applications etc.   **chris** is part of that team.
There's also a "**compute2**" team/scope with a **jack** user in it.

There is also a team called **marketdata** with a scope and streams into which stock market data was ingested etc. 

The desire is that members from compute1 and compute2 can read any streams (present and future) from marketdata but not alter any data. Additionally not be able to mess with reader groups synchronizer states in marketdata or reader groups from the other team's scope.

ACLs would look like this:

```
READ on prn://scope:marketdata/* and READ_UPDATE on prn://scope:compute1/* to chris
READ on prn://scope:marketdata/* and READ_UPDATE on prn://scope:compute1/* to jack
```

chris can read marketata/streamXYZ and is tracking its RG state in compute1/...
jack can read marketata/streamXYZ and is tracking its RG state in compute2/...

It is assumed that if chris's friend maddy, with common READ_UPDATE rights on compute1 enters the picture, she can see and update the RG that chris created.  But she's part of the same team.  That is ok).  jack however, is prevented from accessing chris and maddy's RGs, while still benefiting from accessing marketdata's streams.

If the READ on prn://scope:marketdata/* was allowing mutation of reader groups, then in theory chris and jack could create reader groups there and mess with each other's reader groups. So in this use case, it's best to remain with the "Mutation/access to Reader groups is still protected by READ_UPDATE", such that chris and jack are forced to put their reader groups in their own team scopes.

The flag that's added in this PR, when set to false, allows the above to work.

Reply:
@sarlaccpit thanks for patiently explaining the reasoning behind the option. If a user is granted permission to read from a stream, then it should be automatically granted permission to write to the corresponding reader group stream because without that it can't be part of the reader group. Anything that is needed for a user to be able to read from a stream from an authorization perspective needs to be included in the read authorization, otherwise, from a usability perspective, it won't look very good. The option we are discussing is fairly obscure.

The use of wildcards you describe makes sense. If users with read authorization can create reader groups, then they would be messing up with their own reader group. If that's undesirable, which is reasonable, then perhaps we can require read_update to create a reader group, and only read to join.

One point I don't understand: what happens if a reader group is configured to read from stream A and B, and a given reader that tries to join is only authorized to read from A? I suppose it won't be able to join the reader group.

Reply:
The nuances of join versus create, wrt Read group do make sense.   Although that seems like part of what @ravisharda is likely planning in the future; which is the addition of finer grained permissions, beyond READ and READ_UPDATE which admittedly are not fine grain enough (in general that is, not just for reader groups).  One can imagine the need for things like DELETE, and also WRITE_ONLY (append only) that I have seen mentioned in other context.

But I don't think this is part of this PR (which is already ambitious as it is :).   So @ravisharda can correct me, but basically READ_UPDATE on the reader group basically means: use the reader group and do what you need with it.

For the question of the reader group tracking multiple streams, not all of which you may have access to.   Again, @ravisharda  can say best, but I believe the permission on the reader group is what the focus is, the reader group being its own entity.   The fact it's _related_ to streams is sort of an aside I think?  Meaning, at the end of the day, when the client tries to access stream a or stream b, it'll be confronted to whether or not it has access for these streams then.

Reply:
@fpj To help make what @sarlaccpit says above a bit more concrete, let me share a couple of examples from the newly added integration tests: 

The following lines from a test that uses read-only permissions to read data from a stream, shows the low-level permissions - all `READ` permissions - on the stream, on the reader group and on the scope - required for reading from the stream: 

https://github.com/pravega/pravega/blob/11daa62619d48b59369586c0aa8444d7df75359e/test/integration/src/test/java/io/pravega/test/integration/ReadWithReadPermissionsTest.java#L71-L74

The above is true by default. If you turn a flag on (discussed earlier), reading data from a stream will continue to require READ_UPDATE permissions (as they do today): 

https://github.com/pravega/pravega/blob/11daa62619d48b59369586c0aa8444d7df75359e/test/integration/src/test/java/io/pravega/test/integration/ReadWithReadPermissionsTest.java#L88-L93

Hope that clarifies.

Reply:
I believe I understand what the two of you are saying, but why do we need the option? Said differently, if we remove the configuration option and have only the behavior of requiring read permission to write to internal streams, then what bad things are going to happen?

Reply:
The initial idea of this PR was to require read permissions for internal streams. But then @sarlaccpit explained a situation where that could be problematic for the platform where this feature would be used. He has explained it in his comment above - copying a fragment for reference:

> The desire is that members from compute1 and compute2 can read any streams (present and future) from marketdata but not alter any data. Additionally not be able to mess with reader groups synchronizer states in marketdata or reader groups from the other team's scope.

Reply:
Just adding a quick nuance on "_situation where that could be problematic for the platform where this feature would be used_".

I'd say it'd be problematic for any environments that use the pravega scope as the main unit of authorization.

Reply:
It does not help much because you are trying to convince me that the behavior you are trying to implement makes sense, while the problem for me is not the behavior you are trying to implement, but instead how this PR is proposing to implement it. 

There are two important points for me:

- In that comment, there is an implicit assumption that the use of the wildcard will lead to all state synchronizer streams and reader groups to be readable. We clearly don't want that. We want that a user that is granted permission to read from a given stream (or set of streams via wildcard) is able to only read from those streams and participate in corresponding reader groups. There is even the question of whether a user authorized to read from a given stream can join any reader group reading from that streams. I'd claim that it shouldn't be like that.
- I'm not convinced we need this configuration flag.  As I see it, you can implement the behavior you want without having this flag, which I find confusing from the point of view of a user. I don't understand why we would want to give the user the ability to turn it on and off.

Reply:
Regarding the first point: 

> We want that a user that is granted permission to read from a given stream (or set of streams via wildcard) is able to only read from those streams and participate in corresponding reader groups. 

Yes, a subject (a user) can only read from streams that it has read access to. 
Not sure what you mean by "corresponding" reader groups. Reader groups (RG) are independent w.r.t. authorization. The user must have access to te RG stream as well: 
* say `prn::/scope:MarketData/reader-group:PriceChangeCalculator,READ` (by default or if the Controller config `controller.security.auth.internalStreams.writesWithReadPermissions.enable` is true)
*  `prn::/scope:MarketData/reader-group:PriceChangeCalculator,READ_UPDATE` (if the same config is turned off)

Regarding the second point:

> I'm not convinced we need this configuration flag. As I see it, you can implement the behavior you want without having this flag, which I find confusing from the point of view of a user. I don't understand why we would want to give the user the ability to turn it on and off.

It's really an admin-level config, with a sensible default. The default config lets readers read from a stream with read-only permissions (as shown below), while allowing the admins control over what the readers can read: 
`prn::/scope:MarketData,READ;prn::/scope:MarketData/stream:StockPriceUpdates,READ;prn::/scope:MarketData/reader-group:PriceChangeCalculator,READ`

When the flag is turned on the same user requires this ACL to read data: 
`prn::/scope:MarketData,READ_UPDATE;prn::/scope:MarketData/stream:StockPriceUpdates,READ;prn::/scope:MarketData/reader-group:PriceChangeCalculator,READ_UPDATE`

The SDP platform intends to keep the config as false, as they use authorizations at a scope level using wildcards. Here's an example of SDP's usage that this config being `false` will support: 

* User `Bob` has `READ_UPDATE` permission on `BobScope*`. He can create an RG under `BobScope`, create new streams under `BobScope` and read/write to any streams there. 
* User `Alice` has `READ_UPDATE` permission on `AliceScope*`, and she can create an RG under `AliceScope`, create new streams under `AliceScope` and read/write to any streams there. 
* `Alice` also has `READ` permission on `BobScope*`, and therefore her RG `AliceRg` created for her in `AliceScope` can be used to read from `BobScope/BobStream`. Everything until here is possible in both configurations)
* However, since the config is off in this deployment, she cannot create RGs or join any other existing RGs in BobScope. If this config was set as default/true, she could. 

Hope that clarifies.

Reply:
I'm still not seeing a justification for supporting different behavior. The clarification only explains that there is a use case that limits the participation in reader groups for a principal that is authorized to read. It makes sense to limit participation in reader groups, so I'm wondering why we want to support the other behavior.

As for the reader group authorization options:

```
prn::/scope:MarketData/reader-group:PriceChangeCalculator,READ
prn::/scope:MarketData/reader-group:PriceChangeCalculator,READ_UPDATE
```

My confusion comes from the fact that if `READ_UPDATE` is expected, then what does it mean to use `READ` in an ACL for a reader group. The same for the other case. A reader group in principle can't be read-only. A principal is either authorized to join a reader group or not.  If it is authorized, then it must be able to read and write... or, are you thinking that requesting a checkpoint is a read-only activity that can be performed by a principal with read-only access to the reader group?

Reply:
> My confusion comes from the fact that if READ_UPDATE is expected, then what does it mean to use READ in an ACL for a reader group.

Since clients post RG state into the RG internal stream, even just reading data required `READ_UPDATE` permissions on the RG stream. In this PR, that behavior is dependent on whether the `controller.security.auth.internalStreams.writesWithReadPermissions.enable` is on/off. If it is on, just `READ` permission is sufficient, otherwise `READ_UPDATE` is required. 

> A principal is either authorized to join a reader group or not. If it is authorized, then it must be able to read and write... or, are you thinking that requesting a checkpoint is a read-only activity that can be performed by a principal with read-only access to the reader group?

The intent has been to allow RG checkpointing using `read` permissions, while  providing a configuration to disable that globally (for systems that use scope-level authorizations so that admins have greater control, as explained earlier).

Reply:
To add a bit more context, the flag is useful for systems that use Pravega authorizations at the level of Pravega scopes along with wildcards for child resources like resource groups and streams (as explained earlier in the chain). Ideally, we'd want any external authorization policies applied at that level to be maped to finer-grained authorizations applied by Pravcega at the level of Controller and Segment Store operations. 

However, Pravega doesn't support an important feature in this regard: a way of notifying external access control systems (like Keycloak) of creation/updation/deletion of fine-grained resources like streams, reader groups, etc. In the absence of that feature, mapping coarse-grained authorization policies to fine-grained operation-level authorizations is impractical, and that explains the use of wildcards that @sarlaccpit was referring to earliers. I think we should live with this new config  until we have a solution for this problem. Once we have it, we could remove this config later.